stdlib/runtime/runtimeDebug.gr

```diff
@@ -0,0 +1,712 @@
+        print("new object at 0x" ++ toHex(userPtr))
+      }
+
+      // Add the pointer to the new object to start tracking it until it's deallocated.
+      Set.add(userPtr, allocations)
+    }
+  }
+
+  let onFree = (userPtr: Number) => {
+    totalFrees += 1
+
+    let refCount = readRefCount(userPtr)
+    let tag = readTag(userPtr)
+
+    if (Set.contains(userPtr, allocations)) {
+      if (options.printMallocAndFree) {
+        print(
+          "deallocating object at 0x" ++ toHex(userPtr) ++ " (ref. count: " ++
+          toString(refCount) ++
+          "; tag: " ++
+          toString(tag) ++
+          " " ++
+          tagToTypeName(tag) ++
+          ")" ++
+          maybeObjectContentsAppendix(userPtr),
+        )
+      }
+
+      Set.remove(userPtr, allocations)
+
+      incrementCountInMap(tag, mallocPlusFreeByTag)
+    } else {
+      if (options.printMallocAndFree) {
+        print(
+          "deallocating pre-existing object at 0x" ++ toHex(userPtr) ++
+          " (ref. count: " ++
+          toString(refCount) ++
+          "; tag: " ++
+          toString(tag) ++
+          " " ++
+          tagToTypeName(tag) ++
+          ")" ++
+          maybeObjectContentsAppendix(userPtr),
+        )
+      }
+    }
+
+    if (Map.contains(userPtr, incRefsOnPreExistingObjects) ||
+    Map.contains(userPtr, decRefsOnPreExistingObjects)) {
+      // While this isn't guaranteed to be a real problem, it likely is.
+      // So let's keep count of these cases in order to report it in
+      // the summary and return failed result for test assertions.
+      totalPreexistingObjectsDeallocated += 1
+
+      // Must remove this pointer from the two maps, otherwise it would
+      // lead to problems when trying to print the object info at the end.
+      // Also once freed, this address can be reused, which would then
+      // lead to wrong information being reported.
+      Map.remove(userPtr, incRefsOnPreExistingObjects)
+      Map.remove(userPtr, decRefsOnPreExistingObjects)
+      Map.remove(userPtr, inintialRefCountsOnPreExistingObjects)
+
+      print(
+        "⚠ POSSIBLE PROBLEM DETECTED: pre-existing object at 0x" ++
+        toHex(userPtr) ++
+        " has been deallocated" ++
+        " (ref. count: " ++
+        toString(refCount) ++
+        "; tag: " ++
+        toString(tag) ++
+        " " ++
+        tagToTypeName(tag) ++
+        "). Information about incRefs/decRefs won't be printed for this object." ++
+        maybeObjectContentsAppendix(userPtr),
+      )
+    }
+  }
+
+  @disableGC
+  let incRefHook = (userPtr: WasmI32) => {
+    Memory.incRef(WasmI32.fromGrain(Conv.wasmI32ToNumber))
+    let ptrAsNumber = Conv.wasmI32ToNumber(userPtr)
+
+    // Skip the incRef made when calling the actual function to test.
+    if (WasmI32.ne(userPtr, WasmI32.fromGrain(f)) &&
+    WasmI32.ne(userPtr, WasmI32.fromGrain(GC.removeDebugHooks))) {
+      Memory.incRef(WasmI32.fromGrain(onIncRef))
+      onIncRef(ptrAsNumber)
+    }
```

Comment:
We don't need a `decRef(<self>)` postamble at the end of these functions?

Reply:
You're right. Also for some reason I was incRefing them twice from the GC module.

Since I settled for the hook functions to accept pointers as WasmI32, I think it actually makes more sense now to remove the incRefs on them rather than having them decRef themselves.

If the hook functions were to receive Numbers or Int32s then of course it would make sense for them to use proper GC conventions.